<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Get histogram of fragment lengths from a bam file — bin_fragment_size • ctDNAtools</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Get histogram of fragment lengths from a bam file — bin_fragment_size" />
<meta property="og:description" content="The function first extracts fragment length from a bam file then computes the histogram over defined bins. If normalized is TRUE, 
the counts per bin will be normalized to the total read counts. Optionally,
it can computes the histogram of fragment lengths only for mutated reads (confirmed ctDNA molecules)." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctDNAtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ctDNAtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alkodsi/ctDNAtools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get histogram of fragment lengths from a bam file</h1>
    <small class="dont-index">Source: <a href='https://github.com/alkodsi/ctDNAtools/blob/master/R/bin_fragment_size.R'><code>R/bin_fragment_size.R</code></a></small>
    <div class="hidden name"><code>bin_fragment_size.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function first extracts fragment length from a bam file then computes the histogram over defined bins. If normalized is TRUE, 
the counts per bin will be normalized to the total read counts. Optionally,
it can computes the histogram of fragment lengths only for mutated reads (confirmed ctDNA molecules).</p>
    </div>

    <pre class="usage"><span class='fu'>bin_fragment_size</span>(<span class='no'>bam</span>, <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>targets</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tag</span> <span class='kw'>=</span> <span class='st'>""</span>,
  <span class='kw'>bin_size</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>custom_bins</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='no'>F</span>, <span class='kw'>min_size</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>max_size</span> <span class='kw'>=</span> <span class='fl'>400</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bam</th>
      <td><p>path to bam file.</p></td>
    </tr>
    <tr>
      <th>mutations</th>
      <td><p>An optional data frame with mutations. Must have the columns CHROM, POS, REF, ALT.</p></td>
    </tr>
    <tr>
      <th>targets</th>
      <td><p>a data frame with the target regions to restrict the reads in the bam. Must have three columns: chr, start and end</p></td>
    </tr>
    <tr>
      <th>tag</th>
      <td><p>the RG tag if the bam has more than one samplee.</p></td>
    </tr>
    <tr>
      <th>bin_size</th>
      <td><p>the width of the bin (breaks) of the histogram.</p></td>
    </tr>
    <tr>
      <th>custom_bins</th>
      <td><p>A numeric vector for custom breaks to bin the histogram of fragment length. Over-rides bin_size.</p></td>
    </tr>
    <tr>
      <th>normalized</th>
      <td><p>A logical, whether to normalize the counts to the total number of reads.</p></td>
    </tr>
    <tr>
      <th>min_size</th>
      <td><p>Integer with the lowest fragment length.</p></td>
    </tr>
    <tr>
      <th>max_size</th>
      <td><p>Integer with the highest fragment length.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Other parameters passed to get_fragment_size.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame with one column for the used breaks and one having the histogram (normalized) counts. If mutations is supplied, the output will have one breaks column and three columns
corresponding to variant allele reads, reference allele reads, and other reads. 
Each row has the count of fragment lengths within the bin and optionally normalized by the total number of reads.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Fragment length will extracted from the bam file according to the parameters passed to <code>get_fragment_size</code>, and histogram counts (optionally normalized to total counts)
are computed. Both equal histogram bins via bin_size and manually customized bins via custom_bins are supported.</p>
<p>By using an input mutations, the function will bin separately the reads that support variant alleles, reference alleles and other reads.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>get_fragment_size</code> <code>analyze_fragmentation</code> <code>summarize_fragment_size</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>'targets'</span>,<span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'ctDNAtools'</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>'mutations'</span>,<span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'ctDNAtools'</span>)
<span class='no'>bamT1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'T1.bam'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'ctDNAtools'</span>)

<span class='co'>## basic usage</span>
<span class='fu'>bin_fragment_size</span>(<span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>)

<span class='co'>## with normalization</span>
<span class='fu'>bin_fragment_size</span>(<span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>, <span class='kw'>normalized</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)


<span class='co'>## binning reads categorized based on mutations ref and alt</span>
<span class='fu'>bin_fragment_size</span>(<span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>, <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>)

<span class='co'>## Restrict to reads into targets</span>
<span class='fu'>bin_fragment_size</span>(<span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>, <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Amjad Alkodsi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


