<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Filters a set of mutations — filter_mutations • ctDNAtools</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Filters a set of mutations — filter_mutations" />
<meta property="og:description" content="This function Filters a set of mutations given the input black list or the prevalence of their mismatches in a set of bam files. Mutations that
have more than min_alt_reads in more than min_samples will be removed when no black list is given." />
<meta property="og:image" content="https://alkodsi.github.io/ctDNAtools/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctDNAtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ctDNAtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alkodsi/ctDNAtools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filters a set of mutations</h1>
    <small class="dont-index">Source: <a href='https://github.com/alkodsi/ctDNAtools/blob/master/R/filter_mutations.R'><code>R/filter_mutations.R</code></a></small>
    <div class="hidden name"><code>filter_mutations.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function Filters a set of mutations given the input black list or the prevalence of their mismatches in a set of bam files. Mutations that
have more than min_alt_reads in more than min_samples will be removed when no black list is given.</p>
    </div>

    <pre class="usage"><span class='fu'>filter_mutations</span>(<span class='no'>mutations</span>, <span class='kw'>bams</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>tags</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span>(<span class='st'>""</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>bams</span>)), <span class='kw'>min_alt_reads</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>min_samples</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>min_base_quality</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>max_depth</span> <span class='kw'>=</span> <span class='fl'>1e+05</span>, <span class='kw'>min_mapq</span> <span class='kw'>=</span> <span class='fl'>30</span>,
  <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mutations</th>
      <td><p>A data frame with the reporter mutations. Should have the columns CHROM, POS, REF, ALT.</p></td>
    </tr>
    <tr>
      <th>bams</th>
      <td><p>a vector of paths to bam files</p></td>
    </tr>
    <tr>
      <th>black_list</th>
      <td><p>a character vector of genomic loci of format chr_pos to filter. If not given, the bams
will be scanned for mismatches in the mutations loci and the specified thresholds will be applied for filtering.</p></td>
    </tr>
    <tr>
      <th>tags</th>
      <td><p>a vector of the RG tags if the bam has more than one sample</p></td>
    </tr>
    <tr>
      <th>min_alt_reads</th>
      <td><p>the threshold of read counts showing alternative allele for a sample to be counted</p></td>
    </tr>
    <tr>
      <th>min_samples</th>
      <td><p>the threshold of number of samples above which the mutations is filtered</p></td>
    </tr>
    <tr>
      <th>min_base_quality</th>
      <td><p>minimum base quality for a read to be counted</p></td>
    </tr>
    <tr>
      <th>max_depth</th>
      <td><p>maximum depth above which sampling will happen</p></td>
    </tr>
    <tr>
      <th>min_mapq</th>
      <td><p>the minimum mapping quality for a read to be counted</p></td>
    </tr>
    <tr>
      <th>substitution_specific</th>
      <td><p>logical, whether to have the loci of black_list by substitutions.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a named list contains:</p><ul>
<li><p>ref: vector of read counts of the reference alleles</p></li>
<li><p>alt: vector of read counts of the alternative allele</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Filter a set of mutations using one of two options:</p>
<dl class='dl-horizontal'>
<dt>1.</dt><dd><p>By providing a black list (recommended), which includes a vector of genomic loci chr_pos when substitution_specific is false,
   or chr_pos_ref_alt when substitutions_specific is true. In this mode, all mutations reported in the black list are simply removed.</p></dd>

<dt>2.</dt><dd><p>By providing a set of bam files. The function will run a similar functionality to <code>create_background_panel</code> and filter
   mutations based on the min_alt_reads and min_samples criteria.</p></dd>

</dl>

<p>This function is called internally in <code>test_ctDNA</code> so you likely won't need to use it yourself.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>create_black_list</code> <code>test_ctDNA</code> <code>create_background_panel</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"mutations"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='fu'>filter_mutations</span>(<span class='no'>mutations</span>, <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='st'>"chr14_106327474_C_G"</span>)</div><div class='output co'>#&gt; <span class='message'>Filtering mutations ...</span></div><div class='output co'>#&gt; <span class='message'>Dropped 1 mutations</span></div><div class='output co'>#&gt;    CHROM       POS REF ALT       PHASING
#&gt; 2  chr14 106327649   G   T          &lt;NA&gt;
#&gt; 3  chr14 106327759   A   T          &lt;NA&gt;
#&gt; 4  chr14 106327821   T   C          &lt;NA&gt;
#&gt; 5  chr14 106327838   T   A          &lt;NA&gt;
#&gt; 6  chr14 106327869   C   A 106327869_C_A
#&gt; 7  chr14 106327884   A   C 106327869_C_A
#&gt; 8  chr14 106327909   A   C 106327869_C_A
#&gt; 9  chr14 106327929   A   G 106327869_C_A
#&gt; 10 chr14 106327966   C   T          &lt;NA&gt;</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Amjad Alkodsi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


