<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute the background mismatch rate from a bam file — get_background_rate • ctDNAtools</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Compute the background mismatch rate from a bam file — get_background_rate" />
<meta property="og:description" content="Runs through the target regions base by base counting the mismatches. Then it divides sum(mismatches)/sum(depths) for all bases in the targets" />
<meta property="og:image" content="https://alkodsi.github.io/ctDNAtools/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctDNAtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ctDNAtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alkodsi/ctDNAtools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute the background mismatch rate from a bam file</h1>
    <small class="dont-index">Source: <a href='https://github.com/alkodsi/ctDNAtools/blob/master/R/get_background_rate.R'><code>R/get_background_rate.R</code></a></small>
    <div class="hidden name"><code>get_background_rate.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Runs through the target regions base by base counting the mismatches. Then it divides sum(mismatches)/sum(depths) for all bases in the targets</p>
    </div>

    <pre class="usage"><span class='fu'>get_background_rate</span>(<span class='no'>bam</span>, <span class='no'>targets</span>, <span class='no'>reference</span>, <span class='kw'>vaf_threshold</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>tag</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>min_base_quality</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>max_depth</span> <span class='kw'>=</span> <span class='fl'>1e+05</span>, <span class='kw'>min_mapq</span> <span class='kw'>=</span> <span class='fl'>30</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bam</th>
      <td><p>path to bam file</p></td>
    </tr>
    <tr>
      <th>targets</th>
      <td><p>a data frame with the target regions. Must have three columns: chr, start and end</p></td>
    </tr>
    <tr>
      <th>reference</th>
      <td><p>the reference genome in BSgenome format</p></td>
    </tr>
    <tr>
      <th>vaf_threshold</th>
      <td><p>the bases with higher than this VAF threshold will be ignored in the calculation (real mutations)</p></td>
    </tr>
    <tr>
      <th>tag</th>
      <td><p>the RG tag if the bam has more than one sample</p></td>
    </tr>
    <tr>
      <th>black_list</th>
      <td><p>a character vector of genomic loci of format chr_pos if substitution_specific is false, or chr_pos_ref_alt if substitution_specific is true.
The background will be computed on the target regions after excluding black_list loci.</p></td>
    </tr>
    <tr>
      <th>substitution_specific</th>
      <td><p>logical, whether to have the loci of black_list by substitutions.</p></td>
    </tr>
    <tr>
      <th>min_base_quality</th>
      <td><p>minimum base quality for a read to be counted</p></td>
    </tr>
    <tr>
      <th>max_depth</th>
      <td><p>maximum depth above which sampling will happen</p></td>
    </tr>
    <tr>
      <th>min_mapq</th>
      <td><p>the minimum mapping quality for a read to be counted</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a list containing the general mismatch rate and substitution-specific rates</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Computes the background rate of the input bam file for all bases in the specified targets. Substitutions-specific rates are also calculated.</p>
<p>Genomic positions having non-reference allele frequency higher than vaf_threshold will be excluded (to exclude SNPs and real mutations).</p>
<p>If a black_list is specified, the positions in the black_list (whether substitution_specific or not) will be excluded before computing the background rate.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>create_black_list</code> <code>test_ctDNA</code> <code>create_background_panel</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'>## Load example data</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"targets"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamT1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"T1.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamN1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"N1.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamN2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"N2.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamN3</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"N3.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)

<span class='co'>## Use human reference genome from BSgenome.Hsapiens.UCSC.hg19 library</span>
<span class='fu'><a href='https://rdrr.io/r/base/message.html'>suppressMessages</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>))

<span class='co'>## basic usage</span>
<span class='fu'>get_background_rate</span>(<span class='no'>bamT1</span>, <span class='no'>targets</span>, <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>)

<span class='co'>## more options</span>
<span class='fu'>get_background_rate</span>(<span class='no'>bamT1</span>, <span class='no'>targets</span>, <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>min_base_quality</span> <span class='kw'>=</span> <span class='fl'>30</span>, <span class='kw'>min_mapq</span> <span class='kw'>=</span> <span class='fl'>40</span>, <span class='kw'>vaf_threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>
)

<span class='co'>## with blacklist</span>
<span class='no'>bg_panel</span> <span class='kw'>&lt;-</span> <span class='fu'>create_background_panel</span>(
  <span class='kw'>bam_list</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>bamN1</span>, <span class='no'>bamN2</span>, <span class='no'>bamN3</span>),
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)

<span class='no'>bl2</span> <span class='kw'>&lt;-</span> <span class='fu'>create_black_list</span>(<span class='no'>bg_panel</span>,
  <span class='kw'>mean_vaf_quantile</span> <span class='kw'>=</span> <span class='fl'>0.99</span>,
  <span class='kw'>min_samples_one_read</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>min_samples_two_reads</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>get_background_rate</span>(<span class='no'>bamT1</span>, <span class='no'>targets</span>, <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='no'>bl2</span>
)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Amjad Alkodsi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


