<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Collapses mutations in phase into one event — merge_mutations_in_phase • ctDNAtools</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Collapses mutations in phase into one event — merge_mutations_in_phase" />
<meta property="og:description" content="Given a mutations data frame and a bam file, this function collapses mutations in phase identified by the ID_column into one event.
While doing that, it ignores the reads that support both the reference and alternative alleles for different mutations in phase." />
<meta property="og:image" content="https://alkodsi.github.io/ctDNAtools/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctDNAtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ctDNAtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alkodsi/ctDNAtools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Collapses mutations in phase into one event</h1>
    <small class="dont-index">Source: <a href='https://github.com/alkodsi/ctDNAtools/blob/master/R/merge_mutations_in_phase.R'><code>R/merge_mutations_in_phase.R</code></a></small>
    <div class="hidden name"><code>merge_mutations_in_phase.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a mutations data frame and a bam file, this function collapses mutations in phase identified by the ID_column into one event.
While doing that, it ignores the reads that support both the reference and alternative alleles for different mutations in phase.</p>
    </div>

    <pre class="usage"><span class='fu'>merge_mutations_in_phase</span>(<span class='no'>mutations</span>, <span class='no'>bam</span>, <span class='kw'>tag</span> <span class='kw'>=</span> <span class='st'>""</span>,
  <span class='kw'>ID_column</span> <span class='kw'>=</span> <span class='st'>"phasingID"</span>, <span class='kw'>min_base_quality</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>min_mapq</span> <span class='kw'>=</span> <span class='fl'>30</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mutations</th>
      <td><p>A data frame with the reporter mutations. Should have the columns CHROM, POS, REF, ALT.</p></td>
    </tr>
    <tr>
      <th>bam</th>
      <td><p>path to bam file</p></td>
    </tr>
    <tr>
      <th>tag</th>
      <td><p>the RG tag if the bam has more than one sample</p></td>
    </tr>
    <tr>
      <th>ID_column</th>
      <td><p>The name of the column in mutations data.frame that has the IDs for mutations in phase.
NA values will be filled automatically by unique mutation identifiers.</p></td>
    </tr>
    <tr>
      <th>min_base_quality</th>
      <td><p>minimum base quality for a read to be counted</p></td>
    </tr>
    <tr>
      <th>min_mapq</th>
      <td><p>integer specifying the minimum mapping quality for reads to be included.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with the following slots:</p><dl class='dl-horizontal'>
 <dt>out:</dt><dd><p>A data frame that has the columns:
      <ul>
<li><p>Phasing_id: the ID of the mutations/event.</p></li>
<li><p>ref: number of reference reads.</p></li>
<li><p>alt: number of alternative reads.</p></li>
<li><p>n_reads_multi_mutation: Number of reads that span more than one mutation in phase.</p></li>
<li><p>all_reads: total number of reads.</p></li>
<li><p>multi_support: number of reads that support the alt allele of multiple mutations in phase.</p></li>
</ul></p></dd>
 <dt>purification_prob:</dt><dd><p>Probability of purification: sum(n_reads_multi_mutation)/sum(all_reads)</p></dd>
 <dt>multi_support:</dt><dd><p>Number of multi-support reads in all mutations/events</p></dd>
 <dt>informative_reads:</dt><dd><p>Number of unique reads covering the mutations/events</p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Mutations in phase are those that are supported by the same reads (same allele). The function doesn't identify mutations in phase, but rather use
an ID column in the input whose name is specified by ID_column to tell which mutations are in phase.</p>
<p>Since two or more mutations can be supported by the same evidence,
this function merges these mutations into one event. The function will also remove the mismatches that are exhibited in only one of the mutations in phase (since
this function is developed for the intent of minimal residual disease testing).</p>
<p>The output will include the merged mutations, the probability of purification, which is defined as the number of 
reads covering at least two mutations in phase devided by the number of informative reads. 
Informative reads count is the total number of unique reads mapping to the mutations input
(including both mutations in phase and other mutations).</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>test_ctDNA</code> <code>get_mutations_read_names</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>'mutations'</span>,<span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'ctDNAtools'</span>)
<span class='no'>bamT1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'T1.bam'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'ctDNAtools'</span>)
<span class='fu'>merge_mutations_in_phase</span>(<span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>[<span class='fl'>5</span>:<span class='fl'>10</span>,], <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>, <span class='kw'>ID_column</span> <span class='kw'>=</span> <span class='st'>"PHASING"</span>)</div><div class='output co'>#&gt; $out
#&gt;            Phasing_id ref alt n_reads_multi_mutation all_reads multi_support
#&gt; 1 chr14:106327838_T_A 359   0                      0       359             0
#&gt; 2       106327869_C_A 469   1                    407       470             1
#&gt; 3 chr14:106327966_C_T 319   2                      0       321             0
#&gt; 
#&gt; $purification_prob
#&gt; [1] 0.353913
#&gt; 
#&gt; $multi_support
#&gt; [1] 1
#&gt; 
#&gt; $informative_reads
#&gt; [1] 610
#&gt; </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Amjad Alkodsi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


