<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Tests the ctDNA positivity of a sample — test_ctDNA • ctDNAtools</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Tests the ctDNA positivity of a sample — test_ctDNA" />
<meta property="og:description" content="Given a set of reporter mutation, this functions counts the reads matching the reporter mutations in the sample to be tested,
estimates the mismatch rate for the sample to be tested, and then runs a Monte Carlo simulation test to determine whether the tested sample is positive or negative." />
<meta property="og:image" content="https://alkodsi.github.io/ctDNAtools/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctDNAtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ctDNAtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/alkodsi/ctDNAtools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Tests the ctDNA positivity of a sample</h1>
    <small class="dont-index">Source: <a href='https://github.com/alkodsi/ctDNAtools/blob/master/R/test_ctDNA.R'><code>R/test_ctDNA.R</code></a></small>
    <div class="hidden name"><code>test_ctDNA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a set of reporter mutation, this functions counts the reads matching the reporter mutations in the sample to be tested,
estimates the mismatch rate for the sample to be tested, and then runs a Monte Carlo simulation test to determine whether the tested sample is positive or negative.</p>
    </div>

    <pre class="usage"><span class='fu'>test_ctDNA</span>(<span class='no'>mutations</span>, <span class='no'>bam</span>, <span class='no'>targets</span>, <span class='no'>reference</span>, <span class='kw'>tag</span> <span class='kw'>=</span> <span class='st'>""</span>,
  <span class='kw'>ID_column</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>vaf_threshold</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>min_base_quality</span> <span class='kw'>=</span> <span class='fl'>30</span>, <span class='kw'>max_depth</span> <span class='kw'>=</span> <span class='fl'>1e+05</span>,
  <span class='kw'>min_mapq</span> <span class='kw'>=</span> <span class='fl'>40</span>, <span class='kw'>bam_list</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>bam_list_tags</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span>(<span class='st'>""</span>,
  <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>bam_list</span>)), <span class='kw'>min_alt_reads</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>min_samples</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/Round.html'>ceiling</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>bam_list</span>)/<span class='fl'>10</span>), <span class='kw'>n_simulations</span> <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>pvalue_threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>123</span>,
  <span class='kw'>informative_reads_threshold</span> <span class='kw'>=</span> <span class='fl'>10000</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>mutations</th>
      <td><p>A data frame with the reporter mutations. Should have the columns CHROM, POS, REF, ALT.</p></td>
    </tr>
    <tr>
      <th>bam</th>
      <td><p>path to bam file</p></td>
    </tr>
    <tr>
      <th>targets</th>
      <td><p>a data frame with the target regions. Must have three columns: chr, start and end</p></td>
    </tr>
    <tr>
      <th>reference</th>
      <td><p>the reference genome in BSgenome format</p></td>
    </tr>
    <tr>
      <th>tag</th>
      <td><p>the RG tag if the bam has more than one sample</p></td>
    </tr>
    <tr>
      <th>ID_column</th>
      <td><p>The name of the column that contains the ID of mutations in phase. All mutations in Phase should have the same ID in that column.</p></td>
    </tr>
    <tr>
      <th>black_list</th>
      <td><p>a character vector of genomic loci to filter. The format is chr_pos if substitution_specific is false or
chr_pos_ref_alt if substitution_specific is true. If given, will override <code>bam_list</code>.</p></td>
    </tr>
    <tr>
      <th>substitution_specific</th>
      <td><p>logical, whether to have the loci of black_list by substitutions.</p></td>
    </tr>
    <tr>
      <th>vaf_threshold</th>
      <td><p>the bases with higher than this VAF threshold will be ignored in the calculation (real mutations)</p></td>
    </tr>
    <tr>
      <th>min_base_quality</th>
      <td><p>minimum base quality for a read to be counted</p></td>
    </tr>
    <tr>
      <th>max_depth</th>
      <td><p>maximum depth above which sampling will happen</p></td>
    </tr>
    <tr>
      <th>min_mapq</th>
      <td><p>the minimum mapping quality for a read to be counted</p></td>
    </tr>
    <tr>
      <th>bam_list</th>
      <td><p>A vector containing the paths to bam files used to filter mutations.
Mutations that have more than min_alt_reads in more than min_samples will be filtered. Using black_list is more recommended.</p></td>
    </tr>
    <tr>
      <th>bam_list_tags</th>
      <td><p>the RG tags for the bams included in bams list.</p></td>
    </tr>
    <tr>
      <th>min_alt_reads</th>
      <td><p>When bam_list is provided, this sets the minimum number of alternative allele reads for a sample to be counted.</p></td>
    </tr>
    <tr>
      <th>min_samples</th>
      <td><p>When number of samples having more than <code>min_alt_reads</code> exceeds this number, the mutation will be filtered.</p></td>
    </tr>
    <tr>
      <th>n_simulations</th>
      <td><p>the number of simulations.</p></td>
    </tr>
    <tr>
      <th>pvalue_threshold</th>
      <td><p>the p-value threshold used to decide positivity or negativity.</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>the random seed.</p></td>
    </tr>
    <tr>
      <th>informative_reads_threshold</th>
      <td><p>the number of informative reads (unique reads mapping to specified mutations) under which the test will be undetermined.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a data frame with the following columns:</p><ul>
<li><p>sample: The sample name taken from SM field in the bam file or file base name</p></li>
<li><p>n_mutations: The number of mutations used in the test.</p></li>
<li><p>n_nonzero_alt: Number of mutations that have at least one read supporting alternative allele.</p></li>
<li><p>total_alt_reads: Total number of reads supporting alternative alleles of all mutations in input.</p></li>
<li><p>mutations_filtered: The number of filtered mutations.</p></li>
<li><p>background_rate: The background rate of the tested sample (after all adjustments)</p></li>
<li><p>informative_reads: The number of unique reads covering the mutations used.</p></li>
<li><p>multi_support_reads: The number of reads that support more than one mutations in phase. Non-zero values is a sign of
   positivity not used in the p-value calculation.</p></li>
<li><p>pvalue: The empirical p-value from the Monte Carlo test.</p></li>
<li><p>decision: The decision can be positive, negative or undetermined.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is the main function to test minimal residual disease by ctDNA positivity in a follow-up sample (e.g. after treatment).
The inputs include a bam file for the follow-up sample to be tested, a list of reporter mutations (detected for example before treatment in a ctDNA positive sample),
and an optional black_list (recommended to use) computed from a list of bam files of healthy-like samples or bam_list of the bam_files to use instead of black_list.</p>
<p>The workflow includes the following steps:</p>
<dl class='dl-horizontal'>
<dt>1.</dt><dd><p>Filtering mutations (optional but recommended): The mutations in the input will be filtered removing the ones reported in the black list. If bam_list is provided,
   the mutations will be filtered according to the min_samples and min_alt_reads parameters. See <code>filter_mutations</code>.</p></dd>

<dt>2.</dt><dd><p>The background rate will be computed for the input bam. The black list will be plugged in to exclude the black-listed loci when computing the background rate.
   The black list can be substitution_specific or not, but in both cases, the background rate will be adjusted accordingly. See <code>get_background_rate</code>.</p></dd>

<dt>3.</dt><dd><p>Counting reference and alternative alleles of the reporter mutations in the bam file.</p></dd>

<dt>4.</dt><dd><p>Merging mutations in phase (optional but recommended): If the ID_column is specified in the mutations input, mutations with the same ID (in phase) will be merged.
   While doing so, it is expected that real traces of mutations will be exhibited in the mutations in phase, while artifacts will manifest only in one of the mutations
   in phase. Therefore, the mismatches that map only to one of the mutations in phase but not the others (the others are covered and show reference allele) will be removed.
   This will lead to reduction of the observed mismatches, and therefore, the computed background rate will be adjusted accordingly:
   new rate = old rate * (1 - purification_probability). See <code>merge_mutations_in_phase</code>.</p></dd>

<dt>5.</dt><dd><p>Monte Carlo test: this approach was used by Newman et al., Nature Biotechnology 2016.
   <ul>
<li><p>Given \(N\) reporter mutations each with depth D<sub>i</sub>, randomly sample variant allele reads X<sub>i</sub>
   under the background rate \(p\) using binomial distribution X<sub>i</sub> ~ Binom(D<sub>i</sub>, \(p\)).</p></li>
<li><p>Repeat <code>n_simuations</code> times.</p></li>
<li><p>Count the number of simulations, where simulated data equal or exceed observed data in jointly two measurements: (1) the average VAF for the \(N\) mutations,
     and (2) the number of mutations with non-zero VAF.</p></li>
<li><p>Compute an empirical p-value as (#successes + 1)/(#simulations + 1)</p></li>
</ul></p></dd>

<dt>6.</dt><dd><p>Make a decision: If number of informative reads is lower than <code>informative_reads_threshold</code> parameter,
 the decision will be undetermined. Otherwise, the <code>pvalue_threshold</code>
   parameters will be used to determine positivity or negativity.</p></dd>

</dl>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>get_background_rate</code> <code>merge_mutations_in_phase</code> <code>create_black_list</code> <code>create_background_panel</code>
   <code>filter_mutations</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Load example data</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"mutations"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"targets"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamT1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"T1.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamT2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"T2.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamN1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"N1.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamN2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"N2.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)
<span class='no'>bamN3</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"N3.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"ctDNAtools"</span>)

<span class='co'>## Use human reference genome from BSgenome.Hsapiens.UCSC.hg19 library</span>
<span class='fu'><a href='https://rdrr.io/r/base/message.html'>suppressMessages</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>))
<span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'>## basic usage</span>
<span class='fu'>test_ctDNA</span>(
  <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>,
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>n_simulation</span> <span class='kw'>=</span> <span class='fl'>100</span>
)

<span class='co'>## More options</span>
<span class='fu'>test_ctDNA</span>(
  <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>,
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>n_simulation</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>min_base_quality</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>min_mapq</span> <span class='kw'>=</span> <span class='fl'>30</span>,
  <span class='kw'>vaf_threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>
)

<span class='co'>## Use phasing information</span>
<span class='fu'>test_ctDNA</span>(
  <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT2</span>,
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>n_simulation</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ID_column</span> <span class='kw'>=</span> <span class='st'>"PHASING"</span>
)

<span class='co'>## Use a black list based on loci</span>
<span class='no'>bg_panel</span> <span class='kw'>&lt;-</span> <span class='fu'>create_background_panel</span>(
  <span class='kw'>bam_list</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>bamN1</span>, <span class='no'>bamN2</span>, <span class='no'>bamN3</span>),
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)

<span class='no'>bl1</span> <span class='kw'>&lt;-</span> <span class='fu'>create_black_list</span>(<span class='no'>bg_panel</span>,
  <span class='kw'>mean_vaf_quantile</span> <span class='kw'>=</span> <span class='fl'>0.99</span>,
  <span class='kw'>min_samples_one_read</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>min_samples_two_reads</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>test_ctDNA</span>(
  <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>,
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>n_simulation</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ID_column</span> <span class='kw'>=</span> <span class='st'>"PHASING"</span>, <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='no'>bl1</span>,
  <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)
}
<span class='co'>## Use a substitution-specific black list</span>
<span class='no'>bg_panel</span> <span class='kw'>&lt;-</span> <span class='fu'>create_background_panel</span>(
  <span class='kw'>bam_list</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='no'>bamN1</span>, <span class='no'>bamN2</span>, <span class='no'>bamN3</span>),
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘purrr’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:XVector’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    compact</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:GenomicRanges’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    reduce</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:IRanges’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    reduce</span></div><div class='input'>
<span class='no'>bl2</span> <span class='kw'>&lt;-</span> <span class='fu'>create_black_list</span>(<span class='no'>bg_panel</span>,
  <span class='kw'>mean_vaf_quantile</span> <span class='kw'>=</span> <span class='fl'>0.99</span>,
  <span class='kw'>min_samples_one_read</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>min_samples_two_reads</span> <span class='kw'>=</span> <span class='fl'>1</span>
)</div><div class='output co'>#&gt; <span class='message'>19 loci added satisfying Mean VAF condition</span></div><div class='output co'>#&gt; <span class='message'>0 loci added satisfying one read condition</span></div><div class='output co'>#&gt; <span class='message'>2 loci added satisfying two reads condition</span></div><div class='output co'>#&gt; <span class='message'>Black list has 19 loci</span></div><div class='input'>
<span class='fu'>test_ctDNA</span>(
  <span class='kw'>mutations</span> <span class='kw'>=</span> <span class='no'>mutations</span>, <span class='kw'>bam</span> <span class='kw'>=</span> <span class='no'>bamT1</span>,
  <span class='kw'>targets</span> <span class='kw'>=</span> <span class='no'>targets</span>, <span class='kw'>reference</span> <span class='kw'>=</span> <span class='no'>BSgenome.Hsapiens.UCSC.hg19</span>,
  <span class='kw'>n_simulation</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>ID_column</span> <span class='kw'>=</span> <span class='st'>"PHASING"</span>, <span class='kw'>black_list</span> <span class='kw'>=</span> <span class='no'>bl2</span>,
  <span class='kw'>substitution_specific</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</div><div class='output co'>#&gt; <span class='message'>Analyzing Sample T1 ...</span></div><div class='output co'>#&gt; <span class='message'>Filtering mutations ...</span></div><div class='output co'>#&gt; <span class='message'>Dropped 0 mutations</span></div><div class='output co'>#&gt; <span class='message'>Estimating background rate ...</span></div><div class='output co'>#&gt; <span class='message'>Getting ref and alt Counts ...</span></div><div class='output co'>#&gt; <span class='message'>merging mutations in phase ...</span></div><div class='output co'>#&gt; <span class='message'>Running Monte Carlo simulations</span></div><div class='output co'>#&gt; <span class='message'>Pvalue =  0.0099009900990099</span></div><div class='output co'>#&gt; <span class='message'>Sample T1 is undetermined</span></div><div class='output co'>#&gt;   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered
#&gt; 1     T1          10             6               9                  0
#&gt;   background_rate informative_reads multi_support_reads     pvalue     decision
#&gt; 1    0.0001111016              1327                   1 0.00990099 undetermined</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Amjad Alkodsi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


