<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of ctDNA sequencing data with ctDNAtools • ctDNAtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of ctDNA sequencing data with ctDNAtools">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctDNAtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ctDNAtools_vignette.html">Analysis of ctDNA sequencing data with ctDNAtools</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/alkodsi/ctDNAtools">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analysis of ctDNA sequencing data with ctDNAtools</h1>
                        <h4 class="author">Amjad Alkodsi</h4>
            
            <h4 class="date">2019-12-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alkodsi/ctDNAtools/blob/master/vignettes/ctDNAtools_vignette.Rmd"><code>vignettes/ctDNAtools_vignette.Rmd</code></a></small>
      <div class="hidden name"><code>ctDNAtools_vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The ctDNAtools package is built on the Rsamtools and the GenomicAlignments R packages providing functionalities to analyze circulating tumor DNA sequencing data. In particular, the ctDNAtools can be used to:</p>
<ol style="list-style-type: decimal">
<li><p>Test minimal residual disease by tracking a set of pre-detected mutations referred to as reporter mutations in a follow-up ctDNA sample.</p></li>
<li><p>Analyze fragmentation patterns, histograms and profiles of cell-free DNA (cfDNA) and ctDNA.</p></li>
</ol>
</div>
<div id="minimal-residual-disease-testing" class="section level1">
<h1 class="hasAnchor">
<a href="#minimal-residual-disease-testing" class="anchor"></a>Minimal residual disease testing</h1>
<p>The amount of ctDNA in plasma can drop dramatically after treatment to a level that makes genomic variants undetectable by conventional variant calling. The functionality of testing minimal residual disease (MRD) aims to track pre-detected mutations (e.g. detected in a pretreatment sample) in a follow-up sample, and determine whether the traces of these mutations can be expected by chance given the background mutation rate (ctDNA negative), or they are significantly higher than the background rate (ctDNA positive). This is implemented in the <code>test_ctDNA</code> function, which uses a set of reporter mutations, a bam file of the sample to be tested, reference genome in BSgenome format, and a target file containing the sequencing targets of the panel. See method details in <code><a href="https://rdrr.io/r/utils/help.html">help(test_ctDNA)</a></code></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ctDNAtools)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(purrr)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyr)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">## </a>
<a class="sourceLine" id="cb1-6" data-line-number="6">## Attaching package: 'dplyr'</a>
<a class="sourceLine" id="cb1-7" data-line-number="7">## The following objects are masked from 'package:stats':</a>
<a class="sourceLine" id="cb1-8" data-line-number="8">## </a>
<a class="sourceLine" id="cb1-9" data-line-number="9">##     filter, lag</a>
<a class="sourceLine" id="cb1-10" data-line-number="10">## The following objects are masked from 'package:base':</a>
<a class="sourceLine" id="cb1-11" data-line-number="11">## </a>
<a class="sourceLine" id="cb1-12" data-line-number="12">##     intersect, setdiff, setequal, union</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb1-14" data-line-number="14"></a>
<a class="sourceLine" id="cb1-15" data-line-number="15">## Load example data</a>
<a class="sourceLine" id="cb1-16" data-line-number="16"></a>
<a class="sourceLine" id="cb1-17" data-line-number="17">## example list of predetected mutations</a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">'mutations'</span>,<span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-19" data-line-number="19">mutations</a>
<a class="sourceLine" id="cb1-20" data-line-number="20">##    CHROM       POS REF ALT       PHASING</a>
<a class="sourceLine" id="cb1-21" data-line-number="21">## 1  chr14 106327474   C   G          &lt;NA&gt;</a>
<a class="sourceLine" id="cb1-22" data-line-number="22">## 2  chr14 106327649   G   T          &lt;NA&gt;</a>
<a class="sourceLine" id="cb1-23" data-line-number="23">## 3  chr14 106327759   A   T          &lt;NA&gt;</a>
<a class="sourceLine" id="cb1-24" data-line-number="24">## 4  chr14 106327821   T   C          &lt;NA&gt;</a>
<a class="sourceLine" id="cb1-25" data-line-number="25">## 5  chr14 106327838   T   A          &lt;NA&gt;</a>
<a class="sourceLine" id="cb1-26" data-line-number="26">## 6  chr14 106327869   C   A 106327869_C_A</a>
<a class="sourceLine" id="cb1-27" data-line-number="27">## 7  chr14 106327884   A   C 106327869_C_A</a>
<a class="sourceLine" id="cb1-28" data-line-number="28">## 8  chr14 106327909   A   C 106327869_C_A</a>
<a class="sourceLine" id="cb1-29" data-line-number="29">## 9  chr14 106327929   A   G 106327869_C_A</a>
<a class="sourceLine" id="cb1-30" data-line-number="30">## 10 chr14 106327966   C   T          &lt;NA&gt;</a>
<a class="sourceLine" id="cb1-31" data-line-number="31"></a>
<a class="sourceLine" id="cb1-32" data-line-number="32">## example target file</a>
<a class="sourceLine" id="cb1-33" data-line-number="33"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">'targets'</span>, <span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-34" data-line-number="34">targets</a>
<a class="sourceLine" id="cb1-35" data-line-number="35">##     chr     start       end</a>
<a class="sourceLine" id="cb1-36" data-line-number="36">## 1 chr14 106327422 106328049</a>
<a class="sourceLine" id="cb1-37" data-line-number="37"></a>
<a class="sourceLine" id="cb1-38" data-line-number="38">## Example bam files for middle-treatment and after-treatment</a>
<a class="sourceLine" id="cb1-39" data-line-number="39">bamT1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'T1.bam'</span>, <span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-40" data-line-number="40">bamT2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'T2.bam'</span>, <span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-41" data-line-number="41"></a>
<a class="sourceLine" id="cb1-42" data-line-number="42">## Eample bam files from samples taken from healthy subjects</a>
<a class="sourceLine" id="cb1-43" data-line-number="43">bamN1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'N1.bam'</span>, <span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-44" data-line-number="44">bamN2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'N2.bam'</span>, <span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-45" data-line-number="45">bamN3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata'</span>, <span class="st">'N3.bam'</span>, <span class="dt">package =</span> <span class="st">'ctDNAtools'</span>)</a>
<a class="sourceLine" id="cb1-46" data-line-number="46"></a>
<a class="sourceLine" id="cb1-47" data-line-number="47">## Reference genome</a>
<a class="sourceLine" id="cb1-48" data-line-number="48"><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BSgenome.Hsapiens.UCSC.hg19))</a></code></pre></div>
<div id="basic-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-usage" class="anchor"></a>Basic usage</h2>
<p>In the basic usage, the read counts for reference and variant alleles of the reporter mutations will be quantified, the background rate of the tested sample will be estimated, and a Monte Carlo based sampling test will determine an emperical p-value. The p-value will only be used to determine positivity if the number of informative reads (number of all unique reads covering the mutations) exceed the specified threshold. Otherwise, the sample will be considered undetermined. Note that the test only accepts single nucleotide variants, and indels are not supported.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">test1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/test_ctDNA.html">test_ctDNA</a></span>(<span class="dt">mutations =</span> mutations,</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">          <span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">          <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19, </a>
<a class="sourceLine" id="cb2-4" data-line-number="4">          <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">          <span class="dt">informative_reads_threshold =</span> <span class="dv">100</span>)</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">## Analyzing Sample T1 ...</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">## Estimating background rate ...</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb2-9" data-line-number="9">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb2-10" data-line-number="10">## Pvalue =  9.99900009999e-05</a>
<a class="sourceLine" id="cb2-11" data-line-number="11">## Sample T1 is positive</a>
<a class="sourceLine" id="cb2-12" data-line-number="12">test1</a>
<a class="sourceLine" id="cb2-13" data-line-number="13">##   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">## 1     T1          10             9              13                  0</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">##   background_rate informative_reads multi_support_reads    pvalue decision</a>
<a class="sourceLine" id="cb2-16" data-line-number="16">## 1    0.0001403503              1327                  NA 9.999e-05 positive</a></code></pre></div>
<p>Several parameters can be adjusted including <code>min_base_quality</code> and <code>min_mapq</code> which specify which reads in the bam file to count.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">test2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/test_ctDNA.html">test_ctDNA</a></span>(<span class="dt">mutations =</span> mutations, </a>
<a class="sourceLine" id="cb3-2" data-line-number="2">          <span class="dt">bam =</span> bamT2,</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">          <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19,</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">          <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">          <span class="dt">informative_reads_threshold =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">          <span class="dt">min_base_quality =</span> <span class="dv">20</span>,</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">          <span class="dt">min_mapq =</span> <span class="dv">30</span>)</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">## Analyzing Sample T2 ...</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">## Estimating background rate ...</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb3-11" data-line-number="11">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb3-12" data-line-number="12">## Pvalue =  1</a>
<a class="sourceLine" id="cb3-13" data-line-number="13">## Sample T2 is negative</a>
<a class="sourceLine" id="cb3-14" data-line-number="14">test2</a>
<a class="sourceLine" id="cb3-15" data-line-number="15">##   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered</a>
<a class="sourceLine" id="cb3-16" data-line-number="16">## 1     T2          10             0               0                  0</a>
<a class="sourceLine" id="cb3-17" data-line-number="17">##   background_rate informative_reads multi_support_reads pvalue decision</a>
<a class="sourceLine" id="cb3-18" data-line-number="18">## 1    0.0001294143              2453                  NA      1 negative</a>
<a class="sourceLine" id="cb3-19" data-line-number="19"></a>
<a class="sourceLine" id="cb3-20" data-line-number="20">## batch runs</a>
<a class="sourceLine" id="cb3-21" data-line-number="21">## use future_map2_dfr for multi-threading</a>
<a class="sourceLine" id="cb3-22" data-line-number="22"></a>
<a class="sourceLine" id="cb3-23" data-line-number="23">tests &lt;-<span class="st"> </span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map2.html">map2_dfr</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(bamT1, bamT2),</a>
<a class="sourceLine" id="cb3-24" data-line-number="24">            <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(mutations, mutations), <span class="co"># in case mutations are different</span></a>
<a class="sourceLine" id="cb3-25" data-line-number="25">            <span class="op">~</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/test_ctDNA.html">test_ctDNA</a></span>(<span class="dt">bam =</span> .x, <span class="dt">mutations =</span> .y,</a>
<a class="sourceLine" id="cb3-26" data-line-number="26">                         <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb3-27" data-line-number="27">                         <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19,</a>
<a class="sourceLine" id="cb3-28" data-line-number="28">                         <span class="dt">informative_reads_threshold =</span> <span class="dv">100</span>))</a>
<a class="sourceLine" id="cb3-29" data-line-number="29">## Analyzing Sample T1 ...</a>
<a class="sourceLine" id="cb3-30" data-line-number="30">## Estimating background rate ...</a>
<a class="sourceLine" id="cb3-31" data-line-number="31">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb3-32" data-line-number="32">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb3-33" data-line-number="33">## Pvalue =  9.99900009999e-05</a>
<a class="sourceLine" id="cb3-34" data-line-number="34">## Sample T1 is positive</a>
<a class="sourceLine" id="cb3-35" data-line-number="35">## Analyzing Sample T2 ...</a>
<a class="sourceLine" id="cb3-36" data-line-number="36">## Estimating background rate ...</a>
<a class="sourceLine" id="cb3-37" data-line-number="37">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb3-38" data-line-number="38">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb3-39" data-line-number="39">## Pvalue =  1</a>
<a class="sourceLine" id="cb3-40" data-line-number="40">## Sample T2 is negative</a>
<a class="sourceLine" id="cb3-41" data-line-number="41">tests</a>
<a class="sourceLine" id="cb3-42" data-line-number="42">##   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered</a>
<a class="sourceLine" id="cb3-43" data-line-number="43">## 1     T1          10             9              13                  0</a>
<a class="sourceLine" id="cb3-44" data-line-number="44">## 2     T2          10             0               0                  0</a>
<a class="sourceLine" id="cb3-45" data-line-number="45">##   background_rate informative_reads multi_support_reads    pvalue decision</a>
<a class="sourceLine" id="cb3-46" data-line-number="46">## 1    0.0001403503              1327                  NA 9.999e-05 positive</a>
<a class="sourceLine" id="cb3-47" data-line-number="47">## 2    0.0001128166              2413                  NA 1.000e+00 negative</a></code></pre></div>
</div>
<div id="using-a-black-list" class="section level2">
<h2 class="hasAnchor">
<a href="#using-a-black-list" class="anchor"></a>Using a black list</h2>
<p>A black list of genomic loci or genomic variants can be constructed from a list of bam files corresponding to samples from healthy subjects. The black list can be plugged in the <code>test_ctDNA</code> function to achieve two goals:</p>
<ol style="list-style-type: decimal">
<li><p>It will filter out variants that are likely false positives, limiting false positive ctDNA tests.</p></li>
<li><p>The black listed (often noisy) loci are excluded when computing the background rate, which lowers the background rate against which the observed traces of mutations are tested, and thereby enhancing sensitivity.</p></li>
</ol>
<p>The ctDNAtools package provides two functions (<code>create_background_panel</code> and <code>create_black_list</code>) to build a black list of genomic loci (chr_pos regardless of substitutions) or variants (chr_pos_ref_alt). Both formats are recognized in <code>test_ctDNA</code> and controlled by the <code>substitution_specific</code> parameter.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">## Black list by loci (substition_specific = FALSE)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">bg_panel1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/create_background_panel.html">create_background_panel</a></span>(<span class="dt">bam_list =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(bamN1, bamN2, bamN3), </a>
<a class="sourceLine" id="cb4-3" data-line-number="3">          <span class="dt">targets =</span> targets, </a>
<a class="sourceLine" id="cb4-4" data-line-number="4">          <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19, </a>
<a class="sourceLine" id="cb4-5" data-line-number="5">          <span class="dt">substitution_specific =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">black_list1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/create_black_list.html">create_black_list</a></span>(bg_panel1,</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">          <span class="dt">mean_vaf_quantile =</span> <span class="fl">0.99</span>,</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">          <span class="dt">min_samples_one_read =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">          <span class="dt">min_samples_two_reads =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">## 7 loci added satisfying Mean VAF condition</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">## 0 loci added satisfying one read condition</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">## 2 loci added satisfying two reads condition</a>
<a class="sourceLine" id="cb4-14" data-line-number="14">## Black list has 7 loci</a>
<a class="sourceLine" id="cb4-15" data-line-number="15"></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(black_list1)</a>
<a class="sourceLine" id="cb4-17" data-line-number="17">## [1] "chr14_106327606" "chr14_106327637" "chr14_106327641" "chr14_106327784"</a>
<a class="sourceLine" id="cb4-18" data-line-number="18">## [5] "chr14_106327814" "chr14_106327923"</a>
<a class="sourceLine" id="cb4-19" data-line-number="19"></a>
<a class="sourceLine" id="cb4-20" data-line-number="20">test3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/test_ctDNA.html">test_ctDNA</a></span>(<span class="dt">mutations =</span> mutations, <span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb4-21" data-line-number="21">          <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19, <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb4-22" data-line-number="22">          <span class="dt">informative_reads_threshold =</span> <span class="dv">100</span>, <span class="dt">black_list =</span> black_list1,</a>
<a class="sourceLine" id="cb4-23" data-line-number="23">          <span class="dt">substitution_specific =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-24" data-line-number="24">## Analyzing Sample T1 ...</a>
<a class="sourceLine" id="cb4-25" data-line-number="25">## Filtering mutations ...</a>
<a class="sourceLine" id="cb4-26" data-line-number="26">## Dropped 0 mutations</a>
<a class="sourceLine" id="cb4-27" data-line-number="27">## Estimating background rate ...</a>
<a class="sourceLine" id="cb4-28" data-line-number="28">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb4-29" data-line-number="29">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb4-30" data-line-number="30">## Pvalue =  9.99900009999e-05</a>
<a class="sourceLine" id="cb4-31" data-line-number="31">## Sample T1 is positive</a>
<a class="sourceLine" id="cb4-32" data-line-number="32">test3</a>
<a class="sourceLine" id="cb4-33" data-line-number="33">##   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered</a>
<a class="sourceLine" id="cb4-34" data-line-number="34">## 1     T1          10             9              13                  0</a>
<a class="sourceLine" id="cb4-35" data-line-number="35">##   background_rate informative_reads multi_support_reads    pvalue decision</a>
<a class="sourceLine" id="cb4-36" data-line-number="36">## 1    0.0001419452              1327                  NA 9.999e-05 positive</a>
<a class="sourceLine" id="cb4-37" data-line-number="37"></a>
<a class="sourceLine" id="cb4-38" data-line-number="38">## Black list by variants (substition_specific = TRUE)</a>
<a class="sourceLine" id="cb4-39" data-line-number="39">bg_panel2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/create_background_panel.html">create_background_panel</a></span>(<span class="dt">bam_list =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(bamN1, bamN2, bamN3),</a>
<a class="sourceLine" id="cb4-40" data-line-number="40">           <span class="dt">targets =</span> targets, </a>
<a class="sourceLine" id="cb4-41" data-line-number="41">           <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19,</a>
<a class="sourceLine" id="cb4-42" data-line-number="42">           <span class="dt">substitution_specific =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-43" data-line-number="43"></a>
<a class="sourceLine" id="cb4-44" data-line-number="44">black_list2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/create_black_list.html">create_black_list</a></span>(bg_panel2,</a>
<a class="sourceLine" id="cb4-45" data-line-number="45">          <span class="dt">mean_vaf_quantile =</span> <span class="fl">0.99</span>,</a>
<a class="sourceLine" id="cb4-46" data-line-number="46">          <span class="dt">min_samples_one_read =</span> <span class="dv">2</span>,</a>
<a class="sourceLine" id="cb4-47" data-line-number="47">          <span class="dt">min_samples_two_reads =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb4-48" data-line-number="48">## 19 loci added satisfying Mean VAF condition</a>
<a class="sourceLine" id="cb4-49" data-line-number="49">## 0 loci added satisfying one read condition</a>
<a class="sourceLine" id="cb4-50" data-line-number="50">## 2 loci added satisfying two reads condition</a>
<a class="sourceLine" id="cb4-51" data-line-number="51">## Black list has 19 loci</a>
<a class="sourceLine" id="cb4-52" data-line-number="52"></a>
<a class="sourceLine" id="cb4-53" data-line-number="53"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(black_list2)</a>
<a class="sourceLine" id="cb4-54" data-line-number="54">## [1] "chr14_106327478_G_C" "chr14_106327494_A_C" "chr14_106327498_A_T"</a>
<a class="sourceLine" id="cb4-55" data-line-number="55">## [4] "chr14_106327546_C_T" "chr14_106327606_C_T" "chr14_106327637_C_T"</a>
<a class="sourceLine" id="cb4-56" data-line-number="56"></a>
<a class="sourceLine" id="cb4-57" data-line-number="57">test4 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/test_ctDNA.html">test_ctDNA</a></span>(<span class="dt">mutations =</span> mutations,</a>
<a class="sourceLine" id="cb4-58" data-line-number="58">          <span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb4-59" data-line-number="59">          <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19, </a>
<a class="sourceLine" id="cb4-60" data-line-number="60">          <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb4-61" data-line-number="61">          <span class="dt">informative_reads_threshold =</span> <span class="dv">100</span>,</a>
<a class="sourceLine" id="cb4-62" data-line-number="62">          <span class="dt">black_list =</span> black_list2,</a>
<a class="sourceLine" id="cb4-63" data-line-number="63">          <span class="dt">substitution_specific =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb4-64" data-line-number="64">## Analyzing Sample T1 ...</a>
<a class="sourceLine" id="cb4-65" data-line-number="65">## Filtering mutations ...</a>
<a class="sourceLine" id="cb4-66" data-line-number="66">## Dropped 0 mutations</a>
<a class="sourceLine" id="cb4-67" data-line-number="67">## Estimating background rate ...</a>
<a class="sourceLine" id="cb4-68" data-line-number="68">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb4-69" data-line-number="69">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb4-70" data-line-number="70">## Pvalue =  9.99900009999e-05</a>
<a class="sourceLine" id="cb4-71" data-line-number="71">## Sample T1 is positive</a>
<a class="sourceLine" id="cb4-72" data-line-number="72"></a>
<a class="sourceLine" id="cb4-73" data-line-number="73">test4</a>
<a class="sourceLine" id="cb4-74" data-line-number="74">##   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered</a>
<a class="sourceLine" id="cb4-75" data-line-number="75">## 1     T1          10             9              13                  0</a>
<a class="sourceLine" id="cb4-76" data-line-number="76">##   background_rate informative_reads multi_support_reads    pvalue decision</a>
<a class="sourceLine" id="cb4-77" data-line-number="77">## 1    0.0001320955              1327                  NA 9.999e-05 positive</a></code></pre></div>
<p>The way how to determine the black list variants or loci is very crucial, so careful selection of the parameters in <code>create_black_list</code> is needed. You can also easily design your own more sophisticated criteria to create a black list from the output of <code>create_background_panel</code>, which reports the depths, variant allele frequency, and number of alternative allele reads for all loci in the targets across the input bam files.</p>
<p>Since <code>create_background_panel</code> can take along time with a large number of bam files, multi-threading is supported. All you have to do is call <code>plan(multiprocess)</code> or other plan from the <code>furrr</code> package before calling <code>create_background_panel</code>.</p>
</div>
<div id="exploiting-variants-in-phase" class="section level2">
<h2 class="hasAnchor">
<a href="#exploiting-variants-in-phase" class="anchor"></a>Exploiting variants in phase</h2>
<p>Variants in phase (aka phased variants) are variants that are exhibited by the same sequencing reads (same allele). They are very common in some types of cancer such as lymphomas, but also can be found in other cancers. Mutect2, for example, reports phased variants in the output vcf (note that they can be collapsed into multiple-nucleotide polymorphisms - MNPs, so make sure to have your variants in un-collapsed format for this function). Phased variants can be useful in MRD because we expect that the real traces of phased variants to show in both or all mutations in phase, whereas an artifact that matches one of the mutations in phase will be only exhibited in one variant.</p>
<p>In the ctDNAtools package, you can supply the column name having an ID column in the <code>mutations</code> input, which contains a common ID for the variants in phase. When provided, the variants in phase will be collapsed and quantified jointly, reads that exhibit only one of the variants but not the others will be purified (mismatches removed), and the background rate will be adjusted according to the level of purification expected. See <code><a href="https://rdrr.io/r/utils/help.html">help(merge_mutations_in_phase)</a></code> and <code><a href="https://rdrr.io/r/utils/help.html">help(test_ctDNA)</a></code> for more details.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">## Exploiting phased variants in the test</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">test5 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/test_ctDNA.html">test_ctDNA</a></span>(<span class="dt">mutations =</span> mutations, </a>
<a class="sourceLine" id="cb5-3" data-line-number="3">          <span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">          <span class="dt">reference =</span> BSgenome.Hsapiens.UCSC.hg19,</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">          <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">          <span class="dt">informative_reads_threshold =</span> <span class="dv">100</span>, </a>
<a class="sourceLine" id="cb5-7" data-line-number="7">          <span class="dt">black_list =</span> black_list2,</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">          <span class="dt">substitution_specific =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb5-9" data-line-number="9">          <span class="dt">ID_column =</span> <span class="st">"PHASING"</span>)</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">## Analyzing Sample T1 ...</a>
<a class="sourceLine" id="cb5-11" data-line-number="11">## Filtering mutations ...</a>
<a class="sourceLine" id="cb5-12" data-line-number="12">## Dropped 0 mutations</a>
<a class="sourceLine" id="cb5-13" data-line-number="13">## Estimating background rate ...</a>
<a class="sourceLine" id="cb5-14" data-line-number="14">## Getting ref and alt Counts ...</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">## merging mutations in phase ...</a>
<a class="sourceLine" id="cb5-16" data-line-number="16">## Running Monte Carlo simulations</a>
<a class="sourceLine" id="cb5-17" data-line-number="17">## Pvalue =  9.99900009999e-05</a>
<a class="sourceLine" id="cb5-18" data-line-number="18">## Sample T1 is positive</a>
<a class="sourceLine" id="cb5-19" data-line-number="19">test5</a>
<a class="sourceLine" id="cb5-20" data-line-number="20">##   sample n_mutations n_nonzero_alt total_alt_reads mutations_filtered</a>
<a class="sourceLine" id="cb5-21" data-line-number="21">## 1     T1          10             6               9                  0</a>
<a class="sourceLine" id="cb5-22" data-line-number="22">##   background_rate informative_reads multi_support_reads    pvalue decision</a>
<a class="sourceLine" id="cb5-23" data-line-number="23">## 1    0.0001111016              1327                   1 9.999e-05 positive</a></code></pre></div>
<p>Notice how using phased variants above led to significant reduction in the background rate.</p>
</div>
<div id="useful-internal-functionalities" class="section level2">
<h2 class="hasAnchor">
<a href="#useful-internal-functionalities" class="anchor"></a>Useful internal functionalities</h2>
<p>The following functions are called internally by <code>test_ctDNA</code>, but are available for usage if you want to build your own framework:</p>
<ol style="list-style-type: decimal">
<li><p><code>get_background_rate</code>: Computing the background rate using a bam file</p></li>
<li><p><code>merge_mutations_in_phase</code>: To merge phased variants in single events, purify mismatches and compute the probability of purification.</p></li>
<li><p><code>get_mutations_read_counts</code>: Useful for forced calling mutations.</p></li>
<li><p><code>get_mutations_read_names</code>: Getting the read IDs covering a list of mutations.</p></li>
</ol>
</div>
</div>
<div id="analysis-of-fragmentation" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis-of-fragmentation" class="anchor"></a>Analysis of fragmentation</h1>
<p>The fragment size distribution and fragmentation patterns of cfDNA and ctDNA is biologically relevant (see this nice <a href="https://www.sciencedirect.com/science/article/pii/S1535610819303861">review</a> by van der Pol and Mouliere to learn more). The ctDNAtools package provides functionalities to analyze fragment size histograms, fragment size profiles, and fragmentation patterns.</p>
<div id="extracting-fragment-size-from-a-bam-file" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-fragment-size-from-a-bam-file" class="anchor"></a>Extracting fragment size from a bam file</h2>
<p>Fragment size is extracted from the isize field in the bam file from informative reads. What makes an informative read is left to the user to customize in the available parameters. See <code><a href="https://rdrr.io/r/utils/help.html">help(get_fragment_size)</a></code> for details.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">fs1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/get_fragment_size.html">get_fragment_size</a></span>(<span class="dt">bam =</span> bamT1, </a>
<a class="sourceLine" id="cb6-3" data-line-number="3">      <span class="dt">mapqFilter =</span> <span class="dv">30</span>, </a>
<a class="sourceLine" id="cb6-4" data-line-number="4">      <span class="dt">isProperPair =</span> <span class="ot">NA</span>, </a>
<a class="sourceLine" id="cb6-5" data-line-number="5">      <span class="dt">min_size =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb6-6" data-line-number="6">      <span class="dt">max_size =</span> <span class="dv">400</span>, </a>
<a class="sourceLine" id="cb6-7" data-line-number="7">      <span class="dt">ignore_trimmed =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb6-8" data-line-number="8">      <span class="dt">simple_cigar =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb6-9" data-line-number="9">      <span class="dt">different_strands =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(fs1)</a>
<a class="sourceLine" id="cb6-11" data-line-number="11">##   Sample                  ID   chr     start       end size</a>
<a class="sourceLine" id="cb6-12" data-line-number="12">## 1     T1 T1_Library1:1956186 chr14 106327322 106327482  161</a>
<a class="sourceLine" id="cb6-13" data-line-number="13">## 2     T1 T1_Library1:1956188 chr14 106327322 106327494  173</a>
<a class="sourceLine" id="cb6-14" data-line-number="14">## 3     T1 T1_Library1:1956200 chr14 106327322 106327557  236</a>
<a class="sourceLine" id="cb6-15" data-line-number="15">## 4     T1 T1_Library1:1956211 chr14 106327327 106327492  166</a>
<a class="sourceLine" id="cb6-16" data-line-number="16">## 5     T1 T1_Library1:1956212 chr14 106327327 106327497  171</a>
<a class="sourceLine" id="cb6-17" data-line-number="17">## 6     T1 T1_Library1:1956221 chr14 106327328 106327614  287</a>
<a class="sourceLine" id="cb6-18" data-line-number="18"></a>
<a class="sourceLine" id="cb6-19" data-line-number="19">## optional target restriction</a>
<a class="sourceLine" id="cb6-20" data-line-number="20">fs2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/get_fragment_size.html">get_fragment_size</a></span>(<span class="dt">bam =</span> bamT1, </a>
<a class="sourceLine" id="cb6-21" data-line-number="21">      <span class="dt">targets =</span> targets)</a>
<a class="sourceLine" id="cb6-22" data-line-number="22"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(fs2)</a>
<a class="sourceLine" id="cb6-23" data-line-number="23">##   Sample                  ID   chr     start       end size</a>
<a class="sourceLine" id="cb6-24" data-line-number="24">## 1     T1 T1_Library1:1956186 chr14 106327322 106327482  161</a>
<a class="sourceLine" id="cb6-25" data-line-number="25">## 2     T1 T1_Library1:1956188 chr14 106327322 106327494  173</a>
<a class="sourceLine" id="cb6-26" data-line-number="26">## 3     T1 T1_Library1:1956200 chr14 106327322 106327557  236</a>
<a class="sourceLine" id="cb6-27" data-line-number="27">## 4     T1 T1_Library1:1956211 chr14 106327327 106327492  166</a>
<a class="sourceLine" id="cb6-28" data-line-number="28">## 5     T1 T1_Library1:1956212 chr14 106327327 106327497  171</a>
<a class="sourceLine" id="cb6-29" data-line-number="29">## 6     T1 T1_Library1:1956221 chr14 106327328 106327614  287</a></code></pre></div>
<p>The start and end in the output are considered the most left and most right coordinate from either reads or mates. Note that the output will contain one row for each read pair satisfying conditions in the bam file. There is also an option to input a list of mutations together with the bam file, which leads to an additional column in the output marking the reads that support alternative alleles.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">fs3 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/get_fragment_size.html">get_fragment_size</a></span>(<span class="dt">bam =</span> bamT1, </a>
<a class="sourceLine" id="cb7-3" data-line-number="3">      <span class="dt">mutations =</span> mutations)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(fs3)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">##   Sample                  ID   chr     start       end size category</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">## 1     T1 T1_Library1:1956186 chr14 106327322 106327482  161      ref</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">## 2     T1 T1_Library1:1956188 chr14 106327322 106327494  173      ref</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">## 3     T1 T1_Library1:1956200 chr14 106327322 106327557  236      ref</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">## 4     T1 T1_Library1:1956211 chr14 106327327 106327492  166      ref</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">## 5     T1 T1_Library1:1956212 chr14 106327327 106327497  171      ref</a>
<a class="sourceLine" id="cb7-11" data-line-number="11">## 6     T1 T1_Library1:1956221 chr14 106327328 106327614  287    other</a></code></pre></div>
</div>
<div id="fragment-size-histograms" class="section level2">
<h2 class="hasAnchor">
<a href="#fragment-size-histograms" class="anchor"></a>Fragment size histograms</h2>
<p>Getting a histogram of the fragment size can be done with the <code>bin_fragment_size</code> function. The function supports fixed bin size or custom bins, and can output the counts or normalized counts in each bin.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">bfs1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/bin_fragment_size.html">bin_fragment_size</a></span>(<span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">          <span class="dt">min_size =</span> <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb8-3" data-line-number="3">          <span class="dt">max_size =</span> <span class="dv">400</span>, </a>
<a class="sourceLine" id="cb8-4" data-line-number="4">          <span class="dt">normalized =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb8-5" data-line-number="5">          <span class="dt">bin_size =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb8-6" data-line-number="6"></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(bfs1)</a>
<a class="sourceLine" id="cb8-8" data-line-number="8">##   Breaks T1</a>
<a class="sourceLine" id="cb8-9" data-line-number="9">## 1    1_6  0</a>
<a class="sourceLine" id="cb8-10" data-line-number="10">## 2   6_11  0</a>
<a class="sourceLine" id="cb8-11" data-line-number="11">## 3  11_16  0</a>
<a class="sourceLine" id="cb8-12" data-line-number="12">## 4  16_21  0</a>
<a class="sourceLine" id="cb8-13" data-line-number="13">## 5  21_26  0</a>
<a class="sourceLine" id="cb8-14" data-line-number="14">## 6  26_31  0</a>
<a class="sourceLine" id="cb8-15" data-line-number="15"></a>
<a class="sourceLine" id="cb8-16" data-line-number="16">## batch execution</a>
<a class="sourceLine" id="cb8-17" data-line-number="17">## you can use multithreading by using</a>
<a class="sourceLine" id="cb8-18" data-line-number="18">## furrr::future_map instead of map</a>
<a class="sourceLine" id="cb8-19" data-line-number="19">bfs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(bamT1, bamT2, bamN1, bamN2, bamN3) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-20" data-line-number="20"><span class="st">  </span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(bin_fragment_size, <span class="dt">bin_size =</span> <span class="dv">5</span>, <span class="dt">normalized =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-21" data-line-number="21"><span class="st">  </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span>(inner_join, <span class="dt">by =</span> <span class="st">"Breaks"</span>)</a>
<a class="sourceLine" id="cb8-22" data-line-number="22"></a>
<a class="sourceLine" id="cb8-23" data-line-number="23"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(bfs)</a>
<a class="sourceLine" id="cb8-24" data-line-number="24">##   Breaks T1 T2 N1 N2 N3</a>
<a class="sourceLine" id="cb8-25" data-line-number="25">## 1    1_6  0  0  0  0  0</a>
<a class="sourceLine" id="cb8-26" data-line-number="26">## 2   6_11  0  0  0  0  0</a>
<a class="sourceLine" id="cb8-27" data-line-number="27">## 3  11_16  0  0  0  0  0</a>
<a class="sourceLine" id="cb8-28" data-line-number="28">## 4  16_21  0  0  0  0  0</a>
<a class="sourceLine" id="cb8-29" data-line-number="29">## 5  21_26  0  0  0  0  0</a>
<a class="sourceLine" id="cb8-30" data-line-number="30">## 6  26_31  0  0  0  0  0</a>
<a class="sourceLine" id="cb8-31" data-line-number="31"></a>
<a class="sourceLine" id="cb8-32" data-line-number="32">bfs <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-33" data-line-number="33"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="dt">cols =</span> <span class="op">-</span>Breaks, <span class="dt">names_to =</span> <span class="st">"Sample"</span>,<span class="dt">values_to =</span> <span class="st">"Counts"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-34" data-line-number="34"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/separate.html">separate</a></span>(Breaks, <span class="dt">into =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"start"</span>, <span class="st">"end"</span>), <span class="dt">sep =</span> <span class="st">"_"</span>, <span class="dt">convert =</span> T) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-35" data-line-number="35"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x  =</span> start, <span class="dt">y =</span> Counts, <span class="dt">color =</span> Sample)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-36" data-line-number="36"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="dt">size =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>()</a></code></pre></div>
<p><img src="ctDNAtools_vignette_files/figure-html/histograms-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">## custom bins</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">bfs2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/bin_fragment_size.html">bin_fragment_size</a></span>(<span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">          <span class="dt">normalized =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">          <span class="dt">custom_bins =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">100</span>,<span class="dv">200</span>),</a>
<a class="sourceLine" id="cb9-7" data-line-number="7">          <span class="dt">min_size =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb9-8" data-line-number="8">          <span class="dt">max_size =</span> <span class="dv">400</span>)</a>
<a class="sourceLine" id="cb9-9" data-line-number="9"></a>
<a class="sourceLine" id="cb9-10" data-line-number="10">bfs2</a>
<a class="sourceLine" id="cb9-11" data-line-number="11">##    Breaks        T1</a>
<a class="sourceLine" id="cb9-12" data-line-number="12">## 1   1_100 0.0000000</a>
<a class="sourceLine" id="cb9-13" data-line-number="13">## 2 100_200 0.7871126</a>
<a class="sourceLine" id="cb9-14" data-line-number="14">## 3 200_400 0.2128874</a>
<a class="sourceLine" id="cb9-15" data-line-number="15"></a>
<a class="sourceLine" id="cb9-16" data-line-number="16">## restricted targets</a>
<a class="sourceLine" id="cb9-17" data-line-number="17">bfs2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/bin_fragment_size.html">bin_fragment_size</a></span>(<span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb9-18" data-line-number="18">          <span class="dt">targets =</span> targets,</a>
<a class="sourceLine" id="cb9-19" data-line-number="19">          <span class="dt">normalized =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb9-20" data-line-number="20">          <span class="dt">custom_bins =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">100</span>,<span class="dv">200</span>),</a>
<a class="sourceLine" id="cb9-21" data-line-number="21">          <span class="dt">min_size =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb9-22" data-line-number="22">          <span class="dt">max_size =</span> <span class="dv">400</span>)</a>
<a class="sourceLine" id="cb9-23" data-line-number="23"></a>
<a class="sourceLine" id="cb9-24" data-line-number="24">bfs2</a>
<a class="sourceLine" id="cb9-25" data-line-number="25">##    Breaks        T1</a>
<a class="sourceLine" id="cb9-26" data-line-number="26">## 1   1_100 0.0000000</a>
<a class="sourceLine" id="cb9-27" data-line-number="27">## 2 100_200 0.7871126</a>
<a class="sourceLine" id="cb9-28" data-line-number="28">## 3 200_400 0.2128874</a></code></pre></div>
</div>
<div id="fragment-size-profiles" class="section level2">
<h2 class="hasAnchor">
<a href="#fragment-size-profiles" class="anchor"></a>Fragment size profiles</h2>
<p>The function <code>summarize_fragment_size</code> provides the functionality to summarize the fragment size of reads in predefined genomic regions. You can use any summary function of your choosing, and several of them simultaneously.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">## create some regions from the targets</a>
<a class="sourceLine" id="cb10-2" data-line-number="2"> regions &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">chr =</span> targets<span class="op">$</span>chr,</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">         <span class="dt">start =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dt">from =</span> targets<span class="op">$</span>start <span class="op">-</span><span class="st"> </span><span class="dv">50</span>, <span class="dt">to =</span> targets<span class="op">$</span>end <span class="op">+</span><span class="st"> </span><span class="dv">50</span>, <span class="dt">by =</span> <span class="dv">30</span>),</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">         <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="st">     </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">end =</span> start <span class="op">+</span><span class="st"> </span><span class="dv">30</span>)</a>
<a class="sourceLine" id="cb10-6" data-line-number="6">     </a>
<a class="sourceLine" id="cb10-7" data-line-number="7">sfs1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/summarize_fragment_size.html">summarize_fragment_size</a></span>(<span class="dt">bam =</span> bamT1,</a>
<a class="sourceLine" id="cb10-8" data-line-number="8">          <span class="dt">regions =</span> regions,</a>
<a class="sourceLine" id="cb10-9" data-line-number="9">          <span class="dt">summary_functions =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">Mean =</span> mean, <span class="dt">SD =</span> sd))</a>
<a class="sourceLine" id="cb10-10" data-line-number="10"></a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(sfs1)</a>
<a class="sourceLine" id="cb10-12" data-line-number="12">##                      Region  T1_Mean    T1_SD</a>
<a class="sourceLine" id="cb10-13" data-line-number="13">## 1 chr14:106327432-106327462 122.0000       NA</a>
<a class="sourceLine" id="cb10-14" data-line-number="14">## 2 chr14:106327462-106327492 150.4286 10.75263</a>
<a class="sourceLine" id="cb10-15" data-line-number="15">## 3 chr14:106327492-106327522 165.1000 13.98361</a>
<a class="sourceLine" id="cb10-16" data-line-number="16">## 4 chr14:106327522-106327552 170.9014 18.02629</a>
<a class="sourceLine" id="cb10-17" data-line-number="17">## 5 chr14:106327552-106327582 174.9853 25.08582</a>
<a class="sourceLine" id="cb10-18" data-line-number="18">## 6 chr14:106327582-106327612 174.5472 23.18700</a>
<a class="sourceLine" id="cb10-19" data-line-number="19"></a>
<a class="sourceLine" id="cb10-20" data-line-number="20">## batch run</a>
<a class="sourceLine" id="cb10-21" data-line-number="21">sfs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(bamT1, bamT2, bamN1, bamN2, bamN3) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb10-22" data-line-number="22"><span class="st">  </span><span class="kw"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(summarize_fragment_size, <span class="dt">regions =</span> regions, <span class="dt">summary_functions =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">Median =</span> median)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb10-23" data-line-number="23"><span class="st">  </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span>(inner_join, <span class="dt">by =</span> <span class="st">"Region"</span>)</a>
<a class="sourceLine" id="cb10-24" data-line-number="24"></a>
<a class="sourceLine" id="cb10-25" data-line-number="25"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(sfs)</a>
<a class="sourceLine" id="cb10-26" data-line-number="26">##                      Region T1_Median T2_Median N1_Median N2_Median N3_Median</a>
<a class="sourceLine" id="cb10-27" data-line-number="27">## 1 chr14:106327462-106327492     152.0     153.0     147.0       155       162</a>
<a class="sourceLine" id="cb10-28" data-line-number="28">## 2 chr14:106327492-106327522     167.0     162.5     162.0       165       151</a>
<a class="sourceLine" id="cb10-29" data-line-number="29">## 3 chr14:106327522-106327552     169.0     169.0     164.0       161       147</a>
<a class="sourceLine" id="cb10-30" data-line-number="30">## 4 chr14:106327552-106327582     170.5     168.0     168.0       174       163</a>
<a class="sourceLine" id="cb10-31" data-line-number="31">## 5 chr14:106327582-106327612     171.0     168.0     174.5       170       168</a>
<a class="sourceLine" id="cb10-32" data-line-number="32">## 6 chr14:106327612-106327642     181.0     173.0     156.0       167       117</a></code></pre></div>
</div>
<div id="fragmentation-patterns" class="section level2">
<h2 class="hasAnchor">
<a href="#fragmentation-patterns" class="anchor"></a>Fragmentation patterns</h2>
<p>Fragmentations patterns in a running genomic tiles (overlapping or non-overlapping bins) within target regions can be computed with the <code>analyze_fragmentation</code> function. It will compute the number of fragment ends, the Windowed Protection Score (WPS), and the total number of fragments in each bin. WPS is defined as the number of fragments completely spanning a window (bin) minus the number of fragments with an endpoint within the same window as reported by <a href="https://www.cell.com/fulltext/S0092-8674(15)01569-X">Snyder et al., Cell 2016</a>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">af &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/ctDNAtools/man/analyze_fragmentation.html">analyze_fragmentation</a></span>(<span class="dt">bam =</span> bamT1, </a>
<a class="sourceLine" id="cb11-2" data-line-number="2">          <span class="dt">targets =</span> targets, </a>
<a class="sourceLine" id="cb11-3" data-line-number="3">          <span class="dt">window_size =</span> <span class="dv">120</span>, </a>
<a class="sourceLine" id="cb11-4" data-line-number="4">          <span class="dt">step_size =</span> <span class="dv">5</span>, </a>
<a class="sourceLine" id="cb11-5" data-line-number="5">          <span class="dt">min_size =</span> <span class="dv">120</span>, </a>
<a class="sourceLine" id="cb11-6" data-line-number="6">          <span class="dt">max_size =</span> <span class="dv">180</span>)</a>
<a class="sourceLine" id="cb11-7" data-line-number="7"></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(af)</a>
<a class="sourceLine" id="cb11-9" data-line-number="9">##     chr     start       end  WPS WPS_adjusted n_fragment_ends</a>
<a class="sourceLine" id="cb11-10" data-line-number="10">## 1 chr14 106327303 106327423 -126   -1.0000000             126</a>
<a class="sourceLine" id="cb11-11" data-line-number="11">## 2 chr14 106327308 106327428 -132   -1.0000000             132</a>
<a class="sourceLine" id="cb11-12" data-line-number="12">## 3 chr14 106327313 106327433 -136   -1.0000000             136</a>
<a class="sourceLine" id="cb11-13" data-line-number="13">## 4 chr14 106327318 106327438 -151   -1.0000000             151</a>
<a class="sourceLine" id="cb11-14" data-line-number="14">## 5 chr14 106327323 106327443 -144   -0.9473684             148</a>
<a class="sourceLine" id="cb11-15" data-line-number="15">## 6 chr14 106327328 106327448 -145   -0.9006211             154</a>
<a class="sourceLine" id="cb11-16" data-line-number="16">##   n_fragment_ends_adjusted n_reads</a>
<a class="sourceLine" id="cb11-17" data-line-number="17">## 1                1.0000000     126</a>
<a class="sourceLine" id="cb11-18" data-line-number="18">## 2                1.0000000     132</a>
<a class="sourceLine" id="cb11-19" data-line-number="19">## 3                1.0000000     136</a>
<a class="sourceLine" id="cb11-20" data-line-number="20">## 4                1.0000000     151</a>
<a class="sourceLine" id="cb11-21" data-line-number="21">## 5                0.9736842     152</a>
<a class="sourceLine" id="cb11-22" data-line-number="22">## 6                0.9565217     161</a>
<a class="sourceLine" id="cb11-23" data-line-number="23"></a>
<a class="sourceLine" id="cb11-24" data-line-number="24">af <span class="op">%&gt;%</span><span class="st"> </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(</a>
<a class="sourceLine" id="cb11-25" data-line-number="25">    <span class="dt">cols =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"WPS_adjusted"</span>, <span class="st">"n_fragment_ends_adjusted"</span>,<span class="st">"n_reads"</span>),</a>
<a class="sourceLine" id="cb11-26" data-line-number="26">    <span class="dt">names_to =</span> <span class="st">"Measurement"</span>,</a>
<a class="sourceLine" id="cb11-27" data-line-number="27">    <span class="dt">values_to =</span> <span class="st">"value"</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb11-28" data-line-number="28"><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">Measurement =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(Measurement, </a>
<a class="sourceLine" id="cb11-29" data-line-number="29">        <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"n_reads"</span>, <span class="st">"n_fragment_ends_adjusted"</span>, <span class="st">"WPS_adjusted"</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb11-30" data-line-number="30"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> start, <span class="dt">y =</span> value, <span class="dt">color =</span> Measurement)) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-31" data-line-number="31"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="dt">size =</span> <span class="dv">1</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-32" data-line-number="32"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>Measurement, <span class="dt">scales =</span> <span class="st">"free"</span>, <span class="dt">nrow =</span> <span class="dv">3</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-33" data-line-number="33"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_linedraw</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb11-34" data-line-number="34"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="dt">x =</span> <span class="st">"Chromosome 14"</span>, <span class="dt">y =</span> <span class="st">""</span>)</a></code></pre></div>
<p><img src="ctDNAtools_vignette_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#minimal-residual-disease-testing">Minimal residual disease testing</a><ul class="nav nav-pills nav-stacked">
<li><a href="#basic-usage">Basic usage</a></li>
      <li><a href="#using-a-black-list">Using a black list</a></li>
      <li><a href="#exploiting-variants-in-phase">Exploiting variants in phase</a></li>
      <li><a href="#useful-internal-functionalities">Useful internal functionalities</a></li>
      </ul>
</li>
      <li>
<a href="#analysis-of-fragmentation">Analysis of fragmentation</a><ul class="nav nav-pills nav-stacked">
<li><a href="#extracting-fragment-size-from-a-bam-file">Extracting fragment size from a bam file</a></li>
      <li><a href="#fragment-size-histograms">Fragment size histograms</a></li>
      <li><a href="#fragment-size-profiles">Fragment size profiles</a></li>
      <li><a href="#fragmentation-patterns">Fragmentation patterns</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Amjad Alkodsi.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
